[Console] Add a Tree Helper + multiple Styles

smnandre opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #... 
| License       | MIT


> [!IMPORTANT] 
> Code examples below are _**not**_ up to date, documentation incoming. 

---

## Feature

The goal of the Tree helper is to ease rendering **hierarchical data** in the console: 
* file system
* configuration
* workflows
* ... 

## Usage

Two methods available, manual or via SymfonyStyle

### SymfonyStyle (easiest)

```php 
$io = new SymfonyStyle($input, $ouput); 

$io->tree([
    'A' => [
        'A1',
        'A2',
    ],
    'B' => [
        'B1' => [
            'B11',
        ],
        'B2',
    ],
]); 
```

Will render

```console
root
├── A
│   ├── A1
│   └── A2
├── B
│   ├── B1
│   │   ├── B11
│   │   └── B12
│   └── B2
└── C
```

You also can pass a custom TreeStyle here, or one of the one availbes (see below)

### Manually

You can build the Tree the way you want thanks to the TreeNode class.

```php
$root = (new TreeNode('root'))
    ->addChild((new TreeNode('A'))
        ->addChild(new TreeNode('A1'))
   // ...

$tree = TreeHelper::create($output, $root, style: $style);
$tree->render();
```

--- 

## Styles

Various styles are available (based on the styles of Table .. and some new ones), or you can also build your own Style from scratch.

### Default

```console
root
├── A
│   ├── A1
│   └── A2
├── B
│   ├── B1
│   │   ├── B11
│   │   └── B12
│   └── B2
└── C
```


### Box

```console
root
┃╸ A
┃  ┃╸ A1
┃  ┗╸ A2
┃╸ B
┃  ┃╸ B1
┃  ┃  ┃╸ B11
┃  ┃  ┗╸ B12
┃  ┗╸ B2
┗╸ C
```        


### Box Double

```console
root
╠═ A
║  ╠═ A1
║  ╚═ A2
╠═ B
║  ╠═ B1
║  ║  ╠═ B11
║  ║  ╚═ B12
║  ╚═ B2
╚═ C
```

### Rounded


```console
root
├─ A
│  ├─ A1
│  ╰─ A2
├─ B
│  ├─ B1
│  │  ├─ B11
│  │  ╰─ B12
│  ╰─ B2
╰─ C
```

### Compact

```console
root
├ A
│ ├ A1
│ └ A2
├ B
│ ├ B1
│ │ ├ B11
│ │ └ B12
│ └ B2
└ C
```

### Light

```console
root
|-- A
|   |-- A1
|   `-- A2
|-- B
|   |-- B1
|   |   |-- B11
|   |   `-- B12
|   `-- B2
`-- C
```


### Minimal


```console
root
 A
.  A1
. . A2
 B
.  B1
. .  B11
. . . B12
. . B2
. C
```

---

smnandre committed:
[Console] Add a Tree Helper + multiple Styles

chalasr commented on src/Symfony/Component/Console/Helper/Tree.php:
```diff
@@ -0,0 +1,120 @@
+/**
+ * @author Simon André <smn.andre@gmail.com>
+ */
+class Tree implements \RecursiveIterator
```
I'd make it final too

chalasr commented on src/Symfony/Component/Console/Helper/Tree.php:
```diff
@@ -0,0 +1,120 @@
+
+    public function hasChildren(): bool
+    {
+        return !empty($this->current()->getChildren());
```
we try to avoid using empty as much as possible

chalasr commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,169 @@
+        return self::style(self::STYLE_ROUNDED);
+    }
+
+    public static function style(string $name): self
```
private maybe?

smnandre replied:
Idea was to allow style selection via "strings". 

... But as I did not used it yet... let's purge what is not required 👍

chalasr reviewed (commented):
I like it!
Memory consumption should be considered as this deals with potentially large directories, ideally nodes should be yielded as they come. 
Regarding the public API, we should remove or internalize any getter/setter for which we don't have a strong use case.

smnandre commented:
> I like it! Memory consumption should be considered as this deals with potentially large directories, ideally nodes should be yielded as they come. 

Yep! I'll make the changes depending on the following questions.

> Regarding the public API, we should remove or internalize any getter/setter for which we don't have a strong use case.

Do you have some examples in mind  ? 


Do we need both `createTree` and `tree` in SymfonyStyle (i used Table as base/inspiration)... 
Not sure there is a real need for the ->tree() method..

chalasr commented:
> Do you have some examples in mind ?

I was refering to `Tree::getStyle()/setStyle()` and `TreeStyle::getPrefix*()`

smnandre commented:
I've removed the getters and added a "appy" method on TreeStyle  (we can consider `RecursiveTreeIterator` is part of the contract?)

```php

    public function applyPrefixes(RecursiveTreeIterator $iterator): void
    {
        $iterator->setPrefixPart(RecursiveTreeIterator::PREFIX_LEFT, $this->prefixLeft);
        $iterator->setPrefixPart(RecursiveTreeIterator::PREFIX_MID_HAS_NEXT, $this->prefixMidHasNext);
        $iterator->setPrefixPart(RecursiveTreeIterator::PREFIX_MID_LAST, $this->prefixMidLast);
        $iterator->setPrefixPart(RecursiveTreeIterator::PREFIX_END_HAS_NEXT, $this->prefixEndHasNext);
        $iterator->setPrefixPart(RecursiveTreeIterator::PREFIX_END_LAST, $this->prefixEndLast);
        $iterator->setPrefixPart(RecursiveTreeIterator::PREFIX_RIGHT, $this->prefixRight);
    }
```

Also removed getStyle / setStyle

Will refactor as lazy iterator but i'm not sure how to handle the cycles yet

chadyred commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,148 @@
+            '╠═ ',
+            '╚═ ',
+            '',
+            '╟─ ',
```
'║ ' ? (https://www.fileformat.info/info/unicode/char/2551/index.htm)

chadyred commented on src/Symfony/Component/Console/Style/SymfonyStyle.php:
```diff
@@ -369,6 +374,29 @@ private function getProgressBar(): ProgressBar
             ?? throw new RuntimeException('The ProgressBar is not started.');
     }
 
+    public function createTree(TreeNode|array|\Iterator|null $node = null): Tree
```
Not tested ? Is it needed for something ? Like createTable ?

smnandre replied:
Exactly... I'll rework it if i find an elegant solution for the generator. It will be much easier with the cycle detection made on render.

chadyred commented on src/Symfony/Component/Console/Tests/Helper/TreeTest.php:
```diff
@@ -0,0 +1,297 @@
+        Root
+        ├── Child 1
+        └── Child 2
+        TREE;
```
To indent like others :)

chadyred commented on src/Symfony/Component/Console/Tests/Helper/TreeStyleTest.php:
```diff
@@ -0,0 +1,205 @@
+        |   |   `-- B12
+        |   `-- B2
+        `-- C
+        TREE;
```
To indent like others :pray:

chadyred commented on src/Symfony/Component/Console/Tests/Helper/TreeStyleTest.php:
```diff
@@ -0,0 +1,205 @@
+        $expected = <<<TREE
+root
+╠═ A
+╟─ ╠═ A1
```
To clean the "frame" style ║ instead of ╟─  if you change it on the other comment

chadyred commented on src/Symfony/Component/Console/Helper/TreeBuilder.php:
```diff
@@ -0,0 +1,75 @@
+/**
+ * @author Simon André <smn.andre@gmail.com>
+ */
+final class TreeBuilder
```
Why not just one class? TreeNode is the 'root,' a self-nested class, and it could be responsible for instantiating itself with things like arrays or whatever. For example, TreeNode::buildYourselfWithAnArray (or just 'fromArray', this is just to illustrate the idea). This would be more fluent and intuitive. WYT ? 

In the TreeNode class :
```php
    public static function fromArray(array $array, ?string $root = null): TreeNode
    {
        $root = new self($root ?? '');
        $root->buildTree($root, $array); // Not static anymore

        return $root;
    }
```

smnandre replied:
I don't know... as you see there is now a fromFinder method... i'm not sure the Tree should know anything about it

chadyred replied:
Yeah, we could keep it to add more "strategy" from anything else that the FInder

chadyred commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,148 @@
+
+    public static function light(): self
+    {
+        return new self(...self::STYLES[self::STYLE_LIGHT]);
```
```suggestion
        return self::style(self::STYLE_LIGHT);
```

chadyred commented on src/Symfony/Component/Console/CHANGELOG.md:
```diff
@@ -4,6 +4,9 @@ CHANGELOG
 
+ * Add `TreeHelper` and `TreeBuilder` to display tree-like structures
+ * Add `TreeStyle` to customize the output of the tree helper
+ * Add `createTree()` and `createTreeBuilder()` methods to `SymfonyStyle`
```
I didn't find the `createTreeBuilder` method, maybe to update as you said it is to have feedback :)

chadyred commented:
@smnandre It will be really useful :pray:

OskarStark commented on src/Symfony/Component/Console/CHANGELOG.md:
```diff
@@ -4,6 +4,9 @@ CHANGELOG
 
+ * Add `TreeHelper` to display tree-like structures
+ * Add `TreeStyle` to customize the output of the tree helper
+ * Add `createTree()` methods to `SymfonyStyle`
```
```suggestion
 * Add `SymfonyStyle::createTree()`
```
?

alexandre-daubois replied:
I wonder if the two first entries should be merged. Usually, we're not that descriptive in the Changelog. Should it be maybe:

```suggestion
 * Add `TreeHelper` and `TreeStyle` to display tree-like structures
 * Add `SymfonyStyle::createTree()`
```

smnandre replied:
Done, thanks!

chadyred commented on src/Symfony/Component/Console/Helper/Tree.php:
```diff
@@ -0,0 +1,118 @@
+    {
+        $treeIterator = new \RecursiveTreeIterator(
+            $this,
+            \RecursiveIteratorIterator::CHILD_FIRST,
```
In the parent RecursiveIteratorIterator the 2nd arg is the "mode" and on its child `RecursiveTreeIterator` it is the 4th.  `RecursiveIteratorIterator::CHILD_FIRST` is a constant for the "mode" but here you use it in the 2nd arg, which is the "flags" argument, this is why "psalm" seems unhappy IMO

chadyred reviewed (commented):
:100:

chadyred commented on src/Symfony/Component/Console/Helper/TreeNode.php:
```diff
@@ -0,0 +1,96 @@
+            $this->children[] = $node;
+        } else {
+            $this->children[] = $node;
+        }
```
Condition block seems not needed ?

chadyred commented on src/Symfony/Component/Console/Helper/TreeNode.php:
```diff
@@ -0,0 +1,96 @@
+    {
+        foreach ($this->children as $child) {
+            if (\is_callable($child)) {
+                yield from $child();
```
```suggestion
                yield from call_user_func($child);
```

stof replied:
why ?

chadyred replied:
I see addChild allow callable and I thought at :

```php
<?php

function testCallableArgs(callable $foo) {
    call_user_func($foo);
}

testCallableArgs(function () { echo "bob"; });

function test2() {
   echo "alice";
}

testCallableArgs('test2');

class Test3 {
    public function test3() {
        echo "bar";
    }
}

testCallableArgs([new Test3, 'test3']);
?>
````

Just use the "()" may not support this cases ? I didn't try

stof replied:
`()` supports those cases (at least in versions of PHP that we care about in Symfony 7.3. The `callable` type was not consistent with the support for the dynamic function call syntax in PHP 7 and older)

chadyred replied:
Oh I just try : it works, and the light was ^^

chadyred commented on src/Symfony/Component/Console/Helper/TreeNode.php:
```diff
@@ -0,0 +1,96 @@
+
+    public function count(): int
+    {
+        $count = 0;
```
```suggestion
        $count = 0;

```

chadyred commented on src/Symfony/Component/Console/Helper/Tree.php:
```diff
@@ -0,0 +1,118 @@
+     */
+    public function render(): void
+    {
+        $treeIterator = new \RecursiveTreeIterator(
+            $this,
+            \RecursiveIteratorIterator::CHILD_FIRST,
+            \RecursiveIteratorIterator::CATCH_GET_CHILD,
+        );
```
```suggestion
        $treeIterator = new \RecursiveTreeIterator(
            $this
        );
```

I think only the first parameter is useful, "childfirst "mode see "child" before other tree node acting as "parent", I think it is not wanted : we want to keep the logical order "parent ->children" 

 [RecursiveIteratorIterator::CATCH_GET_CHILD](https://www.php.net/manual/en/class.recursiveiteratoriterator.php#recursiveiteratoriterator.constants.catch-get-child) will ignore exceptions thrown, same, I don't think it is needed

chadyred commented on src/Symfony/Component/Console/Helper/TreeBuilder.php:
```diff
@@ -0,0 +1,103 @@
+ */
+final class TreeBuilder
+{
+    public static function fromArray(array $array, ?string $root = null): TreeNode
```
```suggestion
    /**
     * Parse an array to build a TreeNode.
     *
     * Usage:
     * <code>
     *   TreeBuilder::fromArray([
     *      'rootKey' => [
     *          'parent' => [
     *               'children => [
     *                   'child1',
     *                   'child2',
     *                ],
     *           ],
     *         'orphan' => 'orphan'
     *    ]
     * ]);
     * </code>
     */
    public static function fromArray(array $array, ?string $root = null): TreeNode
```

When I tried it on a concrete example, I take a moment to look for "how" to map things, maybe provided an example could help a little bit :)

smnandre replied:
Doc will be made for this, yes. But the idea was not to need 'orphan' value here

```php
[
    'a' => [
        'a1',
        'a2'
    ],
    'b',
    'c' => [
        'c1',
        'c2' => [
            'c21' => [
                'c211',
                'c212',
            ],
        ],
    ],
]
```

chadyred commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,126 @@
+            '╰─ ',
+            '',
+            '│  ',
+            '╭─ ',
```
```suggestion
            '   ',
```

chadyred commented on src/Symfony/Component/Console/Tests/Helper/TreeStyleTest.php:
```diff
@@ -0,0 +1,197 @@
+        $root
+            ->addChild((new TreeNode('A'))
+                ->addChild(new TreeNode('A1'))
+                ->addChild(new TreeNode('A2'))
```
```suggestion
                ->addChild((new TreeNode('A2'))
                    ->addChild((new TreeNode('A2.1'))
                        ->addChild(new TreeNode('A2.1.1'))
                        ->addChild(new TreeNode('A2.1.2'))
                    )
                )
```

chadyred commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,126 @@
+            '┗╸ ',
+            '',
+            '┃  ',
+            '┗╸ ',
```
```suggestion
            '   ',
```

chadyred commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,126 @@
+            '╚═ ',
+            '',
+            '║  ',
+            '╚═ ',
```
```suggestion
            '  ',
```

chadyred commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,126 @@
+            '└ ',
+            '',
+            '│ ',
+            '└ ',
```
```suggestion
            '  ',
```

chadyred commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,126 @@
+            '`-- ',
+            '',
+            '|   ',
+            '`-- ',
```
```suggestion
            '    ',
```

chadyred commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,126 @@
+            '. ',
+            '',
+            '. ',
+            '. ',
```
```suggestion
            '  ',
```

chadyred commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,126 @@
+            '└── ',
+            '',
+            '│   ',
+            '└── ',
```
```suggestion
            '    ',
```

chadyred commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,126 @@
+    public static function minimal(): self
+    {
+        return new self(
+            '',
```
```suggestion
            '. ',
```

alexandre-daubois commented on src/Symfony/Component/Console/Helper/Tree.php:
```diff
@@ -0,0 +1,114 @@
+    /**
+     * Traverses the tree with cycle detection.
+     *
+     * @return \Generator<string>
```
I don't think this is right, Jetbrains stubs define the first template argument as TKey and the second one as TYield. You may update it with `\Generator<int, string>` if I get it right (see [the stubs](https://github.com/JetBrains/phpstorm-stubs/blob/03273d078e30d26e5893b49e79bc0bde65585956/standard/_types.php#L156-L163) for ref)

stof replied:
Generator even has 4 template types, not just 2, so `\Generator<int, string>` is also wrong (but we could be using `\Traversable<int, string>` if the calling code only care about iteration)

smnandre commented:
I've fusioned `Tree` and `TreeBuilder` into `TreeHelper`, keeping in line with existing ones.

smnandre commented:
Open to review :)

chalasr commented on src/Symfony/Component/Console/Style/SymfonyStyle.php:
```diff
@@ -369,6 +373,24 @@ private function getProgressBar(): ProgressBar
     }
 
+    /**
+     * @param iterable<string, iterable|string|TreeNode> $nodes
```
Some words or example about what can be passed would be nice either here or in TreeHelper or both

chadyred replied:
HA ! It is to avoid back and forth (with a future link to the doc) :), a sample like :

```php
$treeHelper = new TreeHelper($output, TreeNode::fromValues([
	'a' => [
	    'a1',
	    'a2'
	],
	'b',
	'c' => [
	    'c1',
	    'c2' => [
	        'c21' => [
	            'c211',
	            'c212',
	        ],
	    ],
	],
],
    new TreeNode($rootNode),
), TreeStyle::rounded());

$treeHelper->render();
```

alexandre-daubois commented on src/Symfony/Component/Console/Helper/TreeHelper.php:
```diff
@@ -0,0 +1,145 @@
+
+    private \Iterator $childrenIterator;
+
+    public function __construct(
```
Should this be private if a `create()` method also exists? Having two entrypoints could be confusing, what do you think?

smnandre replied:
Absolutely, thank you!

Refactored into private constructor with readonly ppp & renamed factory `createTree()`  👌

alexandre-daubois commented on src/Symfony/Component/Console/Helper/TreeNode.php:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    /**
+     * @return \Traversable<TreeNode>
```
```suggestion
     * @return \Traversable<int, TreeNode>
```

smnandre replied:
Thank you! 😄

alexandre-daubois commented on src/Symfony/Component/Console/Helper/TreeNode.php:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    /**
+     * @return \Traversable<TreeNode>
```
```suggestion
     * @return \Traversable<int, TreeNode>
```

alexandre-daubois reviewed (approved):
A few comments but looks good otherwise 🙂

nicolas-grekas commented on src/Symfony/Component/Console/Helper/TreeHelper.php:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    /**
+     * @internal
```
I don't think methods should be internal. Why should they be?

smnandre replied:
😶‍🌫️ ux habit  

(removed)

nicolas-grekas commented on src/Symfony/Component/Console/Helper/TreeHelper.php:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @return \Traversable<int, string>
+     */
+    private function traverseWithCycleDetection(\RecursiveTreeIterator $iterator, \SplObjectStorage $visited): \Traversable
```
is there any real benefit to splitting this in a separate method?

smnandre replied:
No more, thank you!

nicolas-grekas commented on src/Symfony/Component/Console/Helper/TreeNode.php:
```diff
@@ -0,0 +1,110 @@
+
+    public function __construct(
+        private readonly string $value = '',
+        ?self $parent = null,
```
do we need this? can't this be moved as an outside concern?

smnandre replied:
😶‍🌫️ ux habit  

(removed)

smnandre replied:
Done, thx

nicolas-grekas commented on src/Symfony/Component/Console/Helper/TreeNode.php:
```diff
@@ -0,0 +1,110 @@
+            $parent->addChild($this);
+        }
+
+        foreach ($children as $child) {
```
iterating in the constructor breaks lazy evaluation of the iterable, is that justified?

nicolas-grekas commented on src/Symfony/Component/Console/Helper/TreeStyle.php:
```diff
@@ -0,0 +1,126 @@
+
+    public static function box(): self
+    {
+        return new self(
```
on one line things would be better visually - not consuming my screen's space for nothing ;)

nicolas-grekas commented on src/Symfony/Component/Console/Style/SymfonyStyle.php:
```diff
@@ -282,7 +285,8 @@ public function createProgressBar(int $max = 0): ProgressBar
      * @param iterable<TKey, TValue> $iterable
-     * @param int|null               $max      Number of steps to complete the bar (0 if indeterminate), if null it will be inferred from $iterable
+     * @param int|null               $max      Number of steps to complete the bar (0 if indeterminate), if null it
+     *                                         will be inferred from $iterable
```
I guess this shouldn't be changed

smnandre replied:
Oops not intentional

chadyred commented on src/Symfony/Component/Console/Helper/TreeNode.php:
```diff
@@ -0,0 +1,105 @@
+        }
+    }
+
+    public static function fromValues(iterable $nodes, ?self $node = null): self
```
s/$node/$parent ?

chadyred commented on src/Symfony/Component/Console/Helper/TreeHelper.php:
```diff
@@ -0,0 +1,111 @@
+        $this->children->rewind();
+    }
+
+    public static function createTree(OutputInterface $output, string|TreeNode|null $root = null, iterable $values = [], ?TreeStyle $style = null): self
```
TreeNode::fromValues has the $values arg before the $node ("parent"), here it the opposite, $root is before, is it wanted ? :)

chadyred commented:
Last comments are none blocking suggestion

fabpot commented:
Thank you @smnandre.
